File: php/tests/WebhooksTest.php
<?php

declare(strict_types=1);

namespace Clarian\Tests;

use Clarian\Webhooks;
use PHPUnit\Framework\TestCase;

final class WebhooksTest extends TestCase
{
    private string $secret = 'whsec_test_secret';
    private string $body;
    private string $ts;

    protected function setUp(): void
    {
        $this->body = '{"id":"evt_1","type":"pix_payin.completed",'
            . '"created_at":"2026-01-01T00:00:00Z","environment":"sandbox",'
            . '"data":{"transaction_id":"ord_abc","status":"completed",'
            . '"amount":19.50,"fee":0.37,"currency":"BRL"}}';
        $this->ts = (new \DateTimeImmutable('now', new \DateTimeZone('UTC')))
            ->format('Y-m-d\TH:i:s.u\Z');
    }

    public function testSignVerifyRoundtrip(): void
    {
        $sig = Webhooks::signPayload($this->secret, $this->ts, $this->body);
        $this->assertTrue(
            Webhooks::verifySignature($this->body, $this->ts, $sig, $this->secret),
        );
    }

    public function testTamperedSignatureRejected(): void
    {
        $sig = Webhooks::signPayload($this->secret, $this->ts, $this->body);
        $tampered = str_replace('ord_abc', 'ord_evil', $this->body);
        $this->assertFalse(
            Webhooks::verifySignature($tampered, $this->ts, $sig, $this->secret),
        );
    }

    public function testWrongSecretRejected(): void
    {
        $sig = Webhooks::signPayload('whsec_wrong', $this->ts, $this->body);
        $this->assertFalse(
            Webhooks::verifySignature($this->body, $this->ts, $sig, $this->secret),
        );
    }

    public function testStaleTimestampRejected(): void
    {
        $stale = (new \DateTimeImmutable('now', new \DateTimeZone('UTC')))
            ->modify('-6 minutes')
            ->format('Y-m-d\TH:i:s.u\Z');
        $sig = Webhooks::signPayload($this->secret, $stale, $this->body);
        $this->assertFalse(
            Webhooks::verifySignature($this->body, $stale, $sig, $this->secret),
        );
    }

    public function testFutureTimestampBeyondToleranceRejected(): void
    {
        $future = (new \DateTimeImmutable('now', new \DateTimeZone('UTC')))
            ->modify('+6 minutes')
            ->format('Y-m-d\TH:i:s.u\Z');
        $sig = Webhooks::signPayload($this->secret, $future, $this->body);
        $this->assertFalse(
            Webhooks::verifySignature($this->body, $future, $sig, $this->secret),
        );
    }

    public function testCustomTolerance(): void
    {
        $old = (new \DateTimeImmutable('now', new \DateTimeZone('UTC')))
            ->modify('-10 seconds')
            ->format('Y-m-d\TH:i:s.u\Z');
        $sig = Webhooks::signPayload($this->secret, $old, $this->body);
        $this->assertFalse(
            Webhooks::verifySignature($this->body, $old, $sig, $this->secret, 5),
        );
        $this->assertTrue(
            Webhooks::verifySignature($this->body, $old, $sig, $this->secret, 60),
        );
    }

    public function testMissingInputsRejected(): void
    {
        $sig = Webhooks::signPayload($this->secret, $this->ts, $this->body);
        $this->assertFalse(Webhooks::verifySignature($this->body, $this->ts, $sig, ''));
        $this->assertFalse(Webhooks::verifySignature($this->body, '', $sig, $this->secret));
        $this->assertFalse(Webhooks::verifySignature($this->body, $this->ts, '', $this->secret));
    }

    public function testExtractHeaders(): void
    {
        $headers = [
            Webhooks::HEADER_TIMESTAMP => $this->ts,
            Webhooks::HEADER_SIGNATURE => 'abc123',
            Webhooks::HEADER_EVENT => 'pix_payin.completed',
            Webhooks::HEADER_DELIVERY_ID => 'del_1',
            Webhooks::HEADER_IDEMPOTENCY_KEY => 'idem_1',
            Webhooks::HEADER_ATTEMPT => '1',
        ];
        $h = Webhooks::extractHeaders($headers);
        $this->assertSame($this->ts, $h['timestamp']);
        $this->assertSame('abc123', $h['signature']);
        $this->assertSame('pix_payin.completed', $h['event']);
        $this->assertSame('del_1', $h['deliveryId']);
        $this->assertSame('idem_1', $h['idempotencyKey']);
        $this->assertSame('1', $h['attempt']);
    }

    public function testExtractHeadersCaseInsensitive(): void
    {
        $h = Webhooks::extractHeaders([
            'x-clarian-timestamp' => 't1',
            'x-clarian-signature' => 's1',
        ]);
        $this->assertSame('t1', $h['timestamp']);
        $this->assertSame('s1', $h['signature']);
    }

    public function testExtractHeadersListValues(): void
    {
        $h = Webhooks::extractHeaders([
            'X-Clarian-Timestamp' => ['t2'],
            'X-Clarian-Signature' => ['s2'],
        ]);
        $this->assertSame('t2', $h['timestamp']);
        $this->assertSame('s2', $h['signature']);
    }

    public function testUnixTimestampAccepted(): void
    {
        $now = (string) time();
        $sig = Webhooks::signPayload($this->secret, $now, $this->body);
        $this->assertTrue(
            Webhooks::verifySignature($this->body, $now, $sig, $this->secret),
        );
    }
}

File: java/src/test/java/finance/clarian/sdk/WebhooksTest.java
package finance.clarian.sdk;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class WebhooksTest {

    private String secret;
    private byte[] body;
    private String ts;

    @BeforeEach
    void setUp() {
        secret = "whsec_test_secret";
        body = (
                "{\"id\":\"evt_1\",\"type\":\"pix_payin.completed\","
                        + "\"created_at\":\"2026-01-01T00:00:00Z\",\"environment\":\"sandbox\","
                        + "\"data\":{\"transaction_id\":\"ord_abc\",\"status\":\"completed\","
                        + "\"amount\":19.50,\"fee\":0.37,\"currency\":\"BRL\"}}"
        ).getBytes(java.nio.charset.StandardCharsets.UTF_8);
        ts = Instant.now().toString();
    }

    @Test
    void signVerifyRoundtrip() {
        String sig = Webhooks.signPayload(secret, ts, body);
        assertTrue(Webhooks.verifySignature(body, ts, sig, secret, Webhooks.DEFAULT_TOLERANCE_SECONDS));
        String sig2 = Webhooks.signPayload(secret, ts, new String(body));
        assertTrue(Webhooks.verifySignature(new String(body), ts, sig2, secret));
    }

    @Test
    void tamperedBodyRejected() {
        String sig = Webhooks.signPayload(secret, ts, body);
        byte[] tampered = new String(body).replace("ord_abc", "ord_evil")
                .getBytes(java.nio.charset.StandardCharsets.UTF_8);
        assertFalse(Webhooks.verifySignature(tampered, ts, sig, secret, Webhooks.DEFAULT_TOLERANCE_SECONDS));
    }

    @Test
    void wrongSecretRejected() {
        String sig = Webhooks.signPayload("whsec_wrong", ts, body);
        assertFalse(Webhooks.verifySignature(body, ts, sig, secret, Webhooks.DEFAULT_TOLERANCE_SECONDS));
    }

    @Test
    void staleTimestampRejected() {
        String stale = Instant.now().minus(6, ChronoUnit.MINUTES).toString();
        String sig = Webhooks.signPayload(secret, stale, body);
        assertFalse(Webhooks.verifySignature(body, stale, sig, secret, Webhooks.DEFAULT_TOLERANCE_SECONDS));
    }

    @Test
    void futureTimestampBeyondToleranceRejected() {
        String future = Instant.now().plus(6, ChronoUnit.MINUTES).toString();
        String sig = Webhooks.signPayload(secret, future, body);
        assertFalse(Webhooks.verifySignature(body, future, sig, secret, Webhooks.DEFAULT_TOLERANCE_SECONDS));
    }

    @Test
    void customTolerance() {
        String old = Instant.now().minus(10, ChronoUnit.SECONDS).toString();
        String sig = Webhooks.signPayload(secret, old, body);
        assertFalse(Webhooks.verifySignature(body, old, sig, secret, 5));
        assertTrue(Webhooks.verifySignature(body, old, sig, secret, 60));
    }

    @Test
    void missingInputsRejected() {
        String sig = Webhooks.signPayload(secret, ts, body);
        assertFalse(Webhooks.verifySignature(body, ts, sig, "", Webhooks.DEFAULT_TOLERANCE_SECONDS));
        assertFalse(Webhooks.verifySignature(body, "", sig, secret, Webhooks.DEFAULT_TOLERANCE_SECONDS));
        assertFalse(Webhooks.verifySignature(body, ts, "", secret, Webhooks.DEFAULT_TOLERANCE_SECONDS));
    }

    @Test
    void extractHeaders() {
        Map<String, String> headers = Map.of(
                Webhooks.HEADER_TIMESTAMP, ts,
                Webhooks.HEADER_SIGNATURE, "abc123",
                Webhooks.HEADER_EVENT, "pix_payin.completed"
        );
        Webhooks.Headers h = Webhooks.extractHeaders(headers);
        assertEquals(ts, h.timestamp());
        assertEquals("abc123", h.signature());
        assertEquals("pix_payin.completed", h.event());
    }

    @Test
    void extractHeadersCaseInsensitive() {
        Webhooks.Headers h = Webhooks.extractHeaders(Map.of(
                "x-clarian-timestamp", "t1",
                "x-clarian-signature", "s1"
        ));
        assertEquals("t1", h.timestamp());
        assertEquals("s1", h.signature());
    }

    @Test
    void extractHeadersListValues() {
        Webhooks.Headers h = Webhooks.extractHeaders(Map.of(
                "X-Clarian-Timestamp", List.of("t2"),
                "X-Clarian-Signature", List.of("s2")
        ));
        assertEquals("t2", h.timestamp());
        assertEquals("s2", h.signature());
    }

    @Test
    void unixTimestampAccepted() {
        String now = String.valueOf(Instant.now().getEpochSecond());
        String sig = Webhooks.signPayload(secret, now, body);
        assertTrue(Webhooks.verifySignature(body, now, sig, secret, Webhooks.DEFAULT_TOLERANCE_SECONDS));
    }
}
